src/reader_disc.cpp

```diff
@@ -0,0 +1,559 @@
+/*
+ *
+ * disc reader algo for DeadDiscReader
+ * Copyright (C) 2007-2009, Gennady "ShultZ" Kozlov <qpxtool@mail.ru>
+ *
+ *
+ * initAllCSSKeys function from libdvdread
+ *
+ */
+
+#define _FILE_OFFSET_BITS 64
+
+#include <cstdio>
+#include <cstdlib>
+
+#if (DVDCSS_KEY_CACHE > 0)
+#include <fcntl.h>
+#include <unistd.h>
+#endif
+
+#include <sys/time.h>
+
+#include "imgwriter.hpp"
+#include "qpx_mmc.hpp"
+#include "sectmap.hpp"
+
+#include "dvd_udf.hpp"
+
+#include "reader.hpp"
+#include "reader_disc.hpp"
+#include "threads.hpp"
+
+#define sector_sz 2048
+
+#if (DVDCSS_KEY_CACHE > 0)
+static int initCSSCache(drive_info *dev) {
+  char psz_buffer[PATH_MAX];
+  char *psz_home;
+  char *psz_cache;
+
+  psz_home = getenv("HOME");
+  if (psz_home == NULL)
+    psz_home = getenv("USERPROFILE");
+
+  /* Cache our keys in ${HOME}/.dvdcss/ */
+  if (psz_home) {
+    snprintf(psz_buffer, PATH_MAX, "%s/.dvdcss", psz_home);
+    psz_buffer[PATH_MAX - 1] = '\0';
+    psz_cache = psz_buffer;
+  }
+
+  if (psz_cache != NULL) {
+    /* Check that we can add the ID directory and the block filename */
+    if (strlen(psz_cache) + 1 + 32 + 1 + (DVD_KEY_SIZE * 2) + 10 + 1 >
+        PATH_MAX) {
+      printf("CSS: cache directory name is too long\n");
+      psz_cache = NULL;
+    }
+  }
+
+  if (!psz_cache)
+    goto nocache;
+
+  /* If the cache is enabled, write the cache directory tag */
+  if (psz_cache) {
+    char *psz_tag =
+        "Signature: 8a477f597d28d172789f06886806bc55\r\n"
+        "# This file is a cache directory tag created by libdvdcss.\r\n"
+        "# For information about cache directory tags, see:\r\n"
+        "#   http://www.brynosaurus.com/cachedir/\r\n";
+    char psz_tagfile[PATH_MAX + 1 + 12 + 1];
+    int i_fd;
+
+    sprintf(psz_tagfile, "%s/CACHEDIR.TAG", psz_cache);
+    i_fd = open(psz_tagfile, O_RDWR | O_CREAT, 0644);
+    if (i_fd >= 0) {
+      write(i_fd, psz_tag, strlen(psz_tag));
```

Comment:
Codacy found an issue: [Does not handle strings that are not \0-terminated; if given one it may perform an over-read (it could cause a crash if unprotected) (CWE-126).  ](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)